packages/fuels-core/src/types/transaction_builders.rs

```diff
@@ -252,48 +245,139 @@ impl ScriptTransactionBuilder {
         ScriptTransactionBuilder {
             gas_price: None,
             gas_limit: None,
+            witness_limit: None,
+            max_fee: None,
             maturity: 0,
             script: vec![],
             script_data: vec![],
             inputs: vec![],
             outputs: vec![],
             witnesses: vec![],
             network_info,
+            gas_estimation_tolerance: 0.05,
             unresolved_signatures: Default::default(),
         }
     }
 
-    fn resolve_fuel_tx(self, base_offset: usize, num_witnesses: u8) -> Result<Script> {
-        let gas_price = self.gas_price.unwrap_or(self.network_info.min_gas_price);
-        let gas_limit = self.gas_limit.unwrap_or(self.network_info.max_gas_per_tx);
+    async fn build_with_provider(self, provider: &impl DryRunner) -> Result<ScriptTransaction> {
+        let is_using_predicates = self.is_using_predicates();
+        let base_offset = if is_using_predicates {
+            self.base_offset()
+        } else {
+            0
+        };
+
+        let num_witnesses = self.num_witnesses()?;
+        let tx = self
+            .resolve_fuel_tx_provider(base_offset, num_witnesses, provider)
+            .await?;
+
+        Ok(ScriptTransaction {
+            tx,
+            is_using_predicates,
+        })
+    }
+
+    fn create_dry_run_witnesses(&self, num_witnesses: u8) -> Vec<Witness> {
+        let unresolved_witnesses_len = self.unresolved_signatures.addr_idx_offset_map.len();
+        repeat(Default::default())
+            // Add one in case there is no witnesses at all
+            .take(num_witnesses as usize + unresolved_witnesses_len + 1)
+            .collect()
+    }
+
+    async fn set_script_gas_limit_to_gas_used(
+        tx: &mut Script,
+        provider: &impl DryRunner,
+        network_info: &NetworkInfo,
+        tolerance: f32,
+    ) -> Result<()> {
+        let consensus_params = &network_info.consensus_parameters;
+        // Add `1` because of rounding
+        let max_gas = tx.max_gas(consensus_params.gas_costs(), consensus_params.fee_params()) + 1;
+
+        // If `gas_limit` was not set use `max_gas_per_tx` but subtract the tx's base maximum cost
+        // TODO: @xgreenx why do I need to / 2
```

Comment:
@xgreenx question for you!

Reply:
I guess because you first need to add an input and only set the gas limit after=)